Fixed #30382 -- Allowed specifying parent classes in force_insert of Model.save(). (#16830)
django/django: docs/ref/models/instances.txt

Passing both parameters is an error: you cannot both insert *and* update at the
same time!

When using :ref:`multi-table inheritance <multi-table-inheritance>`, it's also

> possible to pass the ``force_insert`` with a tuple of parent classes to force
> an ``INSERT`` statements. For example::

Reviewer: ```suggestion
possible to provide a tuple of parents classes to ``force_insert`` in order to force 
`INSERT` statements for each base. For example::
```

    Restaurant(pk=1, name="Bob's Cafe").save(force_insert=(Place,))

    Restaurant(pk=1, name="Bob's Cafe", rating=4).save(force_insert=(Place, Rating))

> You can pass ``force_insert=(models.Model,)`` to force an ``INSERT`` statements
> for all parents. By default, ``force_insert=True`` forces insertion a new row
> only for the current model.

Reviewer: ```suggestion
You can pass ``force_insert=(models.Model,)`` to force an ``INSERT`` statement
for all parents. By default, ``force_insert=True`` only forces the insertion of a new
row for the current model.
```

Passing both parameters is an error: you cannot both insert *and* update at the
same time!

.. versionchanged:: 5.0

    In case of an inherited model the ``force_insert=True`` only forces the
    insertion in the leaf model. ``force_insert=False`` is taken as default for the
    parent models.

    If it is required to force insert the parent models also, then consider passing
    a tuple of Model classes that requires to perform force insertion. For
    example::

>         Child(id=3, name="Child-3", value=10).save(force_insert=(GrandParent, Parent))

Reviewer: I think `(GrandParent, Parent)` is redundant with `GrandParent` as `issubclass(Parent, GrandParent)`?

Author: Yes, that's what I about to tell you as in Django one inherited model class cannot have more than one non-abstract parent class what is the point of passing a tuple, we can pass a single models.Model subclass.

Reviewer: Tuples, are only necessary when dealing with diamond inheritance.

Say that you have

```python
class A(models.Model):
   pass

class B(A):
   pass

class C(A):
   pass

class D(B, C):
   pass
```

It might be interesting to `force_insert` `B` and `C` while not `A`! Only in this case does passing a tuple is necessary as that's only doable by doing `d.save(force_insert=(B, C))`.

Author: Tried this but a_ptr willbe clashing in D, so it won't be a valid  inheritance I think.

Author: `
class GrandParent(models.Model):
    name = models.CharField(max_length=50, null=True)
    value = models.IntegerField(null=True)

class Parent(GrandParent):
    pass

class DiamondParent(GrandParent):
    pin_code = models.CharField(max_length=50, null=True)

class DiamondChild(Parent, DiamondParent):
    pass
`

Reviewer: This is expected, it happens because both intermediary parents must define an explicit `OneToOneField(GrandParent, primary_key=True, parent_link=True)` which distinct names.

    ``force_insert=(models.Model,)`` can be used in order to perform forced
    insertion of all the base classes. For example::

>         Child(id=1, name="Child-1", value=10).save(force_insert=(models.Model,))

Reviewer: Any thoughts about allowing passing a single model class instead of a singleton (e.g. `GrandParent` instead of `(GrandParent,)`).

I seems that it would be coherent with the signature of `issubclass` which allows either a `type` or `tuple[type]` ot be provided. Passing a tuple is also only useful for MTI models that are definitive multiple parents at once which is atypical scenario.

Passing both parameters is an error: you cannot both insert *and* update at the
same time!

.. versionchanged:: 5.0

    In case of an inherited model the ``force_insert=True`` only forces the
    insertion in the leaf model. ``force_insert=False`` is taken as default for
    the parent models.

>     If it is required to force insert the parent models also, then consider
>     passing a ``Model`` class or a tuple of ``Model`` classes that requires
>     to perform force insertion. For example::

Reviewer: ```suggestion
    To force the insertion of parent models as well you can pass the ancestor or a tuple
    of ancestors that should be inserted
```
